proposals/list-patterns.md

```diff
@@ -0,0 +1,93 @@
+
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is [1, 2, 3]` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+### Fixed-length list patterns
+
+The syntax will be modified to include a *list_pattern* defined as below:
+
+```antlr
+primary_pattern
+	: list_pattern
+list_pattern
+	: '[' (pattern (',' pattern)* ','?)? ']'
+```
+
+**Pattern compatibility:** A *list_pattern* is compatible with any type that conforms to the ***range indexer pattern***:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+2. Has an accessible indexer with a single `int` parameter
+3. Has an accessible `Slice` method that takes two `int` parameters (for slice patterns)
+
+This rule includes `T[]`,  `string`,  `Span<T>`, `ImmutableArray<T>` and more.
+
+**Subsumption checking:**  This construct will be built entirely on top of the existing DAG nodes. The subsumption is checked just like recursive patterns with `ITuple` - corresponding subpatterns are matched by position plus an additional node for testing length.
+
+**Lowering:** A pattern of the form `expr is [1, 2, 3]` is equivalent to the following code:
+```cs
+expr.Length is 3
+&& expr[0] is 1
+&& expr[1] is 2
+&& expr[2] is 3
+```
+
+### Variable-length list patterns
+
+The syntax will be modified to include a *slice_pattern* defined as below:
+```antlr
+primary_pattern
+	: slice_pattern
+slice_pattern
+	: `..`
+```
+
+A *slice_pattern* is only permitted once and only directly in a *list_pattern* and discards ***zero or more*** elements. Note that it's possible to use a *slice_pattern* in a nested *list_pattern* e.g. `[.., [.., 1]]` will match `new int[][]{new[]{1}}`.
+
+A *slice_pattern* acts like a proper discard i.e. no tests will be emitted for such pattern, rather it only affects other nodes, namely the length and indexer. For instance, a pattern of the form `expr is [1, .., 3]`  is equivalent to the following code: 
+```cs
+expr.Length is >= 2
+&& expr[0]  is 1
+&& expr[^1] is 3
```

Comment:
Does this represent System.Index, or are you just using this to express `&& expr[expr.Length - 1] is 3`?

Reply:
No Index nor Range would be created except for slice pattern usage on arrays.